---
title: "PlinkQC with LBD CWOW SNP array data"
author: "Kimberly Olney, Ph.D."
date: "07/25/2024"
output:
  html_document:
    theme: cerulean
    toc: true
    toc_float: true
  pdf_document:
    toc: true
editor_options:
  chunk_output_type: inline
---

# PlinkQC 
PlinkQC is a R package for genotype quality control in genetic association studies. See https://meyer-lab-cshl.github.io/plinkQC/index.html

The protocol is implemented in three main functions:\
1) The per-individual quality control (perIndividualQC)\
2) The per-marker quality control (perMarkerQC)\
3) The generation of the new, quality control dataset (cleanData)\

Individuals and markers that fail the quality control can subsequently be removed with plinkQC to generate a new, clean dataset.

Input  CWOW_flipped  712,571
```{r setup, echo=FALSE, warning=FALSE}
knitr::opts_knit$set(root.dir = ".")
knitr::opts_chunk$set(warning = FALSE, message = FALSE, echo=FALSE, results='hide', fig.keep='all')
#library(devtools)
#install_github("meyer-lab-cshl/plinkQC")
library(plinkQC)
library(ggplot2) # save plots via ggsave
Sys.setenv(RSTUDIO_PANDOC="/usr/local/biotools/pandoc/3.1.2/bin")
```

```{r params, echo=FALSE}
# Path to plinkQC package
package.dir <- find.package('plinkQC')
# Path to plink tool 
path2plink <- "../../../../tools/miniconda3/envs/QTL/bin/plink"

# Input directory
indir <- '../snp_array/TOPMED_imput'
# Output directory 
qcdir <- '../snp_array/imputed_post_plinkQC'
# Name of the SNP array input data 
name <- 'final_filtered_data' #  CWOW_n598_GRCh38_updated
```

# Per-individual quality control
PerIndividualQC writes a list of all fail individual IDs to the qcdir. These IDs will be removed in the computation of the perMarkerQC

The HapMapIII and CWOW data were merged prior to running plinkQC. See the README for more information on how the HapMapIII reference data was obtained.
```{r individual_qc, eval=TRUE}
prefixMergedDataset <- c("merge_HAP_CWOW_pca") 

fail_individuals <- perIndividualQC(indir=indir, 
                                    qcdir='../snp_array', 
                                    name=name,
                                    prefixMergedDataset=prefixMergedDataset,
                                    path2plink=path2plink,
                                    europeanTh = 5, # default 1.5 
                                    do.run_check_ancestry = FALSE,
                                    interactive=TRUE, 
                                    verbose=FALSE,
                                    legend_text_size = 10,
                                    legend_title_size = 10,
                                    axis_text_size = 10,
                                    axis_title_size = 10,
                                    title_size = 12)

ggsave("../results/plinkQC_plots/per_individual_check.pdf", width = 8.5, height = 11)
```

# Sex check
check_sex: for the identification of individuals with discordant sex information
```{r sex_check}
# Run sex check 
fail_sex <- check_sex(indir=indir, 
                      qcdir=qcdir, 
                      name=name, 
                      interactive=TRUE,
                      verbose=FALSE, 
                      path2plink=path2plink, 
                      legend_text_size = 10,
                      legend_title_size = 10,
                      axis_text_size = 10,
                      axis_title_size = 10,
                      title_size = 12)

ggsave("../results/plinkQC_plots/sex_check.pdf", width = 6, height = 5)
```

# Heterozygosity check
The identification of individuals with outlying missing genotype and/or heterozygosity rates helps to detect samples with poor DNA quality and/or concentration that should be excluded from the study. Typically, individuals with more than 3-7% of their genotype calls missing are removed. Outlying heterozygosity rates are judged relative to the overall heterozygosity rates in the study, and individuals whose rates are more than a few standard deviations (sd) from the mean heterozygosity rate are removed. A typical quality control for outlying heterozygosity rates would remove individuals who are three sd away from the mean rate.
```{r het_check}
# Run check heterozygosity rate
fail_het_imiss <- check_het_and_miss(indir=indir, 
                                     qcdir=qcdir, 
                                     name=name,
                                     interactive=TRUE,
                                     verbose=FALSE,
                                     path2plink=path2plink,
                                     legend_text_size = 10,
                                     legend_title_size = 10,
                                     axis_text_size = 10,
                                     axis_title_size = 10,
                                     title_size = 12)

ggsave("../results/plinkQC_plots/heterozygosity_check.pdf", width = 7, height = 6)
```

# Relatedness check
Related individuals can be identified by their proportion of shared alleles at the genotyped markers (identity by descend, IBD). Standardly, individuals with second-degree relatedness or higher will be excluded.
```{r related_check}
# Finds pairs of samples whose proportion of IBD is larger than the specified highIBDTh.
# Default highIBDTh = 0.1875,
exclude_relatedness <- check_relatedness(indir=indir, 
                                         qcdir=qcdir, 
                                         name=name,
                                         interactive=TRUE,
                                         verbose=FALSE,
                                         path2plink=path2plink,
                                         legend_text_size = 10,
                                         legend_title_size = 10,
                                         axis_text_size = 10,
                                         axis_title_size = 10,
                                         title_size = 12)

ggsave("../results/plinkQC_plots/IBD_check.pdf", width = 10, height = 6)
```

# Ancestry check
The identification of individuals of divergent ancestry can be achieved by combining the genotypes of the study population with genotypes of a reference dataset consisting of individuals from known ethnicities (for instance individuals from the Hapmap or 1000 genomes study.

Principal component analysis (PCA) on the combined genotype panel is employed to detect population structure, matching the granularity of the reference dataset. The tool check_ancestry is used to identify individuals with divergent ancestry. check_ancestry utilizes information from principal components 1 and 2 to determine the center of the European reference samples. Study samples with a Euclidean distance from this center that exceeds a specified radius are classified as non-European.

The default radius is 1.5. For this dataset, to retain more individuals, a radius of 5 was selected. 
```{r ancestry_check}
# Merged CWOW and HapMap data
prefixMergedDataset <- c("merge_HAP_CWOW_pca") 

# europeanTh - Scaling factor of radius to be drawn around center of European reference sample
# Default 1.5 was changed to 5 

exclude_ancestry <- evaluate_check_ancestry(indir=indir, 
                                            name=name,
                                            europeanTh = 5,
                                            prefixMergedDataset=prefixMergedDataset,
                                            verbose=FALSE,
                                            interactive=TRUE,
                                            legend_text_size = 10,
                                            legend_title_size = 10,
                                            axis_text_size = 10,
                                            axis_title_size = 10,
                                            title_size = 12)

ggsave("../results/plinkQC_plots/ancestry_check.pdf", width = 9, height = 6.5)
```

# Missingness check
Markers with excessive missingness rate are removed as they are considered unreliable. Typically, thresholds for marker exclusion based on missingness range from 1%-5%. Identifying markers with high missingness rates is implemented in snp_missingness. It calculates the rates of missing genotype calls and frequency for all variants in the individuals that passed the perIndividualQC.
```{r snp_miss}
fail_snpmissing <- check_snp_missingness(indir=indir, 
                                         qcdir=qcdir, 
                                         name=name,
                                         interactive=TRUE,
                                         verbose=FALSE,
                                         path2plink=path2plink, 
                                         showPlinkOutput=FALSE,
                                         legend_text_size = 10,
                                         legend_title_size = 10,
                                         axis_text_size = 10,
                                         axis_title_size = 10,
                                         title_size = 12)

ggsave("../results/plinkQC_plots/snp_missingness_check.pdf", width = 10, height = 6)
```

# HWE check 
Markers with strong deviation from HWE might be indicative of genotyping or genotype-calling errors. As serious genotyping errors often yield very low p-values, it is recommended to choose a reasonably low threshold to avoid filtering too many variants (that might have slight, non-critical deviations). Identifying markers with deviation from HWE is implemented in check_hwe. It calculates the observed and expected heterozygote frequencies per SNP in the individuals that passed the perIndividualQC and computes the deviation of the frequencies from Hardy-Weinberg equilibrium (HWE) by HWE exact test.
```{r hwe}
fail_hwe <- check_hwe(indir=indir, 
                      qcdir=qcdir, 
                      name=name, 
                      interactive=TRUE,
                      verbose=FALSE,
                      path2plink=path2plink, 
                      showPlinkOutput=FALSE, 
                      legend_text_size = 10,
                      legend_title_size = 10,
                      axis_text_size = 10,
                      axis_title_size = 10,
                      title_size = 12)
ggsave("../results/plinkQC_plots/hwe_check.pdf", width = 10, height=5)

```

# MAF check 
minor allele frequency (MAF). Markers with low minor allele count are often removed as the actual genotype calling (via the calling algorithm) is very difficult due to the small sizes of the heterozygote and rare-homozygote clusters. check_maf alculates the minor allele frequencies for all variants in the individuals that passed the perIndividualQC.
```{r maf}
fail_maf <- check_maf(indir=indir, 
                      qcdir=qcdir, 
                      name=name, 
                      interactive=TRUE,
                      verbose=FALSE,
                      path2plink=path2plink, 
                      showPlinkOutput=FALSE, 
                      legend_text_size = 10,
                      legend_title_size = 10,
                      axis_text_size = 10,
                      axis_title_size = 10,
                      title_size = 12)
ggsave("../results/plinkQC_plots/maf.pdf", width = 4.5, height=5.5)
```

# Output cleaned data 
```{r clean}
# Overview of failed markers
fail_markers <- perMarkerQC(indir=indir, 
                            qcdir="../snp_array", 
                            name=name,
                            path2plink=path2plink,
                            verbose=TRUE, 
                            interactive=TRUE,
                            showPlinkOutput=FALSE)

# Output cleaned data
Ids  <- cleanData(indir=indir, 
                  qcdir="../snp_array", 
                  name=name, 
                  path2plink=path2plink,
                  verbose=TRUE, 
                  showPlinkOutput=FALSE)
```


