---
title: "Process genotype file"
author: "Kimberly Olney, Ph.D."
date: "07/18/2024"
output:
  html_document:
    df_print: paged
  pdf_document: default
params:
  args: myarg
---

# Setup
```{r setup}
knitr::opts_knit$set(root.dir = ".")
library(dplyr)
library(data.table)
library(edgeR)
# Read in metadata filtered metadata file 
metadata <- read.delim("../metadata/filtered_metadata.txt")
```

# Genotype data set up 
```{r genotype, eval=FALSE}
# Read the raw genotype file
genotype_raw <- fread("../snp_array/Filtered_n598_CWOW_genotype.raw")
rownames(genotype_raw) <- genotype_raw$IID

# Remove the first six columns (FID, IID, PAT, MAT, SEX, PHENOTYPE)
genotype_data <- genotype_raw %>% select(-c(1:6))
rownames(genotype_data) <- genotype_raw$IID
# Transpose the data to get SNPs as rows and samples as columns
genotype_data <- t(genotype_data)
colnames(genotype_data) <- genotype_raw$IID

# Save the formatted data to a file
write.table(genotype_data, file = "../snp_array/Filtered_n598_CWOW_genotype_data.txt", sep = " ", quote = FALSE, col.names = NA)
```

# Counts data set up  
```{r counts}
counts <- fread("../RNA_counts/counts_filtered.txt")
genes <- fread("../RNA_counts/genes_filtered.txt")

y <- DGEList(counts=counts,genes=data.frame(Length=genes$width))
y <- calcNormFactors(y)
RPKM <- rpkm(y)

# Subset the dataframe
counts_filtered <- RPKM[, metadata$NPID, drop = FALSE]
rownames(counts_filtered) <- genes$gene_name

# Create a named vector for matching
matching_vector <- setNames(metadata$IID, metadata$NPID)
# Replace the column names in counts_filtered
colnames(counts_filtered) <- matching_vector[colnames(counts_filtered)]

write.table(counts_filtered, "../RNA_counts/n598_counts_filtered.txt", sep = "\t", quote = FALSE)
```
