---
title: "Run Matrix eQTL"
author: "Kimberly Olney, Ph.D."
date: "March 2025"
output:
  html_document:
    df_print: paged
  pdf_document: default
params:
  args: myarg
---

# Setup
```{r setup}
knitr::opts_knit$set(root.dir = ".")
setwd("/tgen_labs/jfryer/kolney/LBD_CWOW/QTL/LBD_CWOW_QTL/scripts/")
source(here::here("/tgen_labs/jfryer/kolney/LBD_CWOW/QTL/LBD_CWOW_QTL/scripts/", "file_paths_and_colours.R"))
```

# Read inputs for Matrix eQTL 
Inputs include:\
genotype - columns are samples, rows are snps. 0 is homozygous reference, 1 is heterozygous, 2 is homozygous alternate 
expression - columns are samples, rows are genes. Values are TPM expression 
covariates - columns are samples, rows are phenotype information such as sex, age, and RIN
gene info - gene_id, chromosome, start, and end positions
snp annotation - snpids and the corresponding gene_id 
```{r input_for_matrixeqtl}
# Read in gene information. These are the genes used in the bulk RNAseq differential expression analysis.
genes <- fread("../RNA_counts/genes_filtered.txt")
gene_info <- genes[,c(1:4)] # Keep only columns 1 - 4
colnames(gene_info) <- c("geneid","chr", "left", "right") # Rename the columns 
# Read in SNP annotation
# The SNP annotation was download from the illumina website and lifted over to GRCh38
snp_anno <- fread("../snp_array/reference/NCBI_SNP_positions_with_gene_hg19ToHg38_lift_over.tsv", 
                  header = TRUE,  fill = TRUE)

snp_anno <- snp_anno[,c(6,1,2)] # Keep 
colnames(snp_anno) <- c("snpid","chr", "pos") # rename the columns 

# Thresholds, Only associations significant at this level will be saved
pvOutputThreshold = 1e-5
# The p-value threshold determines which gene-SNP associations are saved in the output file output_file_name. 
# For larger datasets the threshold should be lower. 
# Setting the threshold to a high value for a large dataset may cause excessively large output files.
useModel = modelLINEAR # modelANOVA or modelLINEAR or modelLINEAR_CROSS

snps = SlicedData$new()
snps$fileDelimiter = " "      # the TAB character
snps$fileOmitCharacters = "NA" # denote missing values
snps$fileSkipRows = 1          # one row of column labels
snps$fileSkipColumns = 1       # one column of row labels
snps$fileSliceSize = 2000      # read file in pieces of 2,000 rows
snps$LoadFile("../snp_array/Filtered_n580_CWOW_genotype_data_AD_control_imputed.txt")

gene = SlicedData$new()
gene$fileDelimiter = "\t"      # the TAB character
gene$fileOmitCharacters = "NA" # denote missing values
gene$fileSkipRows = 1          # one row of column labels
gene$fileSkipColumns = 1       # one column of row labels
gene$fileSliceSize = 2000      # read file in pieces of 2,000 rows
gene$LoadFile("../RNA_counts/n580_TPM_combat_adjusted_counts_log2_AD_control_imputed.txt")

cvrt = SlicedData$new()
cvrt$fileDelimiter = "\t"      # the TAB character
cvrt$fileOmitCharacters = "NA" # denote missing values
cvrt$fileSkipRows = 1          # one row of column labels
cvrt$fileSkipColumns = 1       # one column of row labels
cvrt$fileSliceSize = 2000      # read file in pieces of 2,000 rows
cvrt$LoadFile("../snp_array/covariates_and_phenotype_files/covariates_AD_control_imputed.txt")
```

# Order check 
Make sure that the gene expression (gene) and genetic datasets (snps) have the samples in the same order!
```{r table_ID_match}
stopifnot(all(colnames(snps) == colnames(gene)))
```

# Multiple testing 
To adjust for the number of tests, we could divide 0.05 by the number of tests to be performed. 
```{r}
snps_rows <- as.numeric(nrow(snps))
gene_rows <- as.numeric(nrow(gene))

# How many tests
snps_rows*gene_rows 

lower.bound <- nrow(snps)
upper.bound <- nrow(snps) * gene_rows
threshold <- 0.05/lower.bound
```


# cis & trans eQTL
Test local and distantance gene-SNP pairs separately 
In case of cis/trans eQTL analysis described below, two output files are produced, one with cis-eQTLs, another only with trans. Every record contains a SNP name, a transcript name, estimate of the effect size, t- or F-statistic, p-value, and FDR.
plot Q-Q plots of local and distant p-values
```{r cis_trans_eqtl}
# Only associations significant at this level will be saved
pvOutputThreshold_cis = 1 # output all cis
pvOutputThreshold_tra = 1e-5

# Distance for local gene-SNP pairs
cisDist = 1e6 # 1 million 
# It means you're looking for SNPs within 1 million base pairs (1 Mb) of a gene's transcription start site (TSS). 
# This threshold has been widely adopted because it captures most regulatory variants in the region while avoiding including too many trans-eQTLs. 
# A balance between capturing local effects and excluding distant regulatory influences.

dis_me <- Matrix_eQTL_main(
    snps = snps,
    gene = gene,
    cvrt = cvrt,
    output_file_name =  file.path("../MatrixEQTL/n579_trans_eQTL_AD_control_imputed"),
    pvOutputThreshold = pvOutputThreshold_tra,
    useModel = modelLINEAR_CROSS, 
    pvOutputThreshold.cis = pvOutputThreshold_cis, # threshold
    output_file_name.cis =  file.path("../MatrixEQTL/n579_cis_eQTL_AD_control_imputed"),
    snpspos = snp_anno, 
    genepos = gene_info[,c("geneid","chr","left","right")],
    cisDist = 1e6, ## define cis as < 1Mb
    verbose = TRUE,
    pvalue.hist = "qqplot",
    min.pv.by.genesnp = FALSE,
    noFDRsaveMemory = FALSE)

## Results:
cat('Analysis done in: ', dis_me$time.in.sec, ' seconds', '\n')
cat('Detected local eQTLs:', '\n')
show(dis_me$cis$eqtls)
cat('Detected distant eQTLs:', '\n')
show(dis_me$trans$eqtls)

## Plot the Q-Q plot of local and distant p-values
#plot(dis_me)
#saveToPDF("../MatrixEQTL/qqplot_pvalues_cis_trans_eqtl.pdf", width = 7, height = 6)

#saveRDS(dis_me, "../MatrixEQTL/n579_cis_and_trans_eQTL.Rds")
```

# Matrix eQTL 
Each significant gene-SNP association is recorded in a separate line in the output file and in the returned object me.
```{r matrixEQTL, eval = FALSE}
me <- Matrix_eQTL_main(
    snps = snps,
    gene = gene,
    cvrt = cvrt,
    output_file_name =  file.path("../MatrixEQTL/n579_all_eQTL_AD_control_imputed"),
    pvOutputThreshold = pvOutputThreshold, 
    useModel = modelLINEAR, ## test using linear models 
    verbose = TRUE,
    pvalue.hist = TRUE,
    min.pv.by.genesnp = FALSE,
    noFDRsaveMemory = FALSE)

## Results:
cat('Analysis done in: ', me$time.in.sec, ' seconds', '\n')
## Plot the histogram of all p-values
plot(me)
# Save as RDS 
saveRDS(me, "../MatrixEQTL/n579_eQTL.rds")
```

