---
title: "Plot eQTLs"
author: "Kimberly Olney, Ph.D."
date: "07/22/2024"
output:
  html_document:
    df_print: paged
  pdf_document: default
params:
  args: myarg
---
# Setup
```{r setup}
knitr::opts_knit$set(root.dir = ".")
library(dplyr)
library(data.table)
library(MatrixEQTL)
library(ggplot2)
```

# Read in MatrixEQTL data
```{r}
me <- fread("../MatrixEQTL/n579_all_eQTL_disease_interaction") # _disease_interaction
genotype_data <- fread("../snp_array/Filtered_n579_CWOW_genotype_data_no_allele_info.txt")
expression_data <- fread("../RNA_counts/n579_TPM_combat_adjusted_counts.txt")
meta <- fread("../metadata/filtered_metadata.txt")
gene_info <- fread("../RNA_counts/genes_filtered.txt")
gene_info$gene <- gene_info$gene_name
```

# Plot top eQTL
```{r plot}
genes <- fread("../RNA_counts/genes_filtered.txt")
me$gene_id <- me$gene
eqtls_gene <- merge(me, genes, by = "gene_id")

idx <- which.min(eqtls_gene$`p-value`)
eqtls_gene[idx,]

top.gene <- as.character(eqtls_gene$gene[idx])
top.snp <- as.character(eqtls_gene$SNP[idx])
top.gene

sliced.data.row <- function(object, row) {
    if (is.factor(row))
        row <- as.character(row)
    if (is.character(row)) {
        row <- which(rownames(object) == row)
        if (length(row) == 0)
            stop("Invalid row")
    }
    slice.size <- nrow(object[[1]])
    slice.idx <- floor(row/slice.size) + 1
    slice <- object[[slice.idx]]
    row <- row - slice.size * (slice.idx - 1)
    ret <- slice[row,]
    names(ret) <- colnames(object)
    ret             
}
```

```{r input_for_matrixeqtl}
# Read in gene information. These are the genes used in the bulk RNAseq differential expression analysis.
genes <- fread("../RNA_counts/genes_filtered.txt")
gene_info <- genes[,c(1:4)] # Keep only columns 1 - 4
colnames(gene_info) <- c("geneid","chr", "left", "right") # Rename the columns 
# Read in SNP annotation
# The SNP annotation was download from the illumina website 
snp_anno <- fread("../snp_array/illumina_OminExpress24/InfiniumOmniExpress-24v1-3_A1.annotated.txt", 
                  header = TRUE,  fill = TRUE)
snp_anno$gene <- sub(",.*", "", snp_anno$`Gene(s)`) # create a gene column 
snp_anno <- snp_anno[,c(1:3)] # Keep only columns 1 - 3
colnames(snp_anno) <- c("snpid","chr", "pos") # rename the columns 

# Thresholds, Only associations significant at this level will be saved
pvOutputThreshold = 1e-8 
# The p-value threshold determines which gene-SNP associations are saved in the output file output_file_name. 
# For larger datasets the threshold should be lower. 
# Setting the threshold to a high value for a large dataset may cause excessively large output files.
useModel = modelLINEAR # modelANOVA or modelLINEAR or modelLINEAR_CROSS

snps = SlicedData$new()
snps$fileDelimiter = " "      # the TAB character
snps$fileOmitCharacters = "NA" # denote missing values
snps$fileSkipRows = 1          # one row of column labels
snps$fileSkipColumns = 1       # one column of row labels
snps$fileSliceSize = 2000      # read file in pieces of 2,000 rows
snps$LoadFile("../snp_array/Filtered_n579_CWOW_genotype_data_no_allele_info.txt")

gene = SlicedData$new()
gene$fileDelimiter = "\t"      # the TAB character
gene$fileOmitCharacters = "NA" # denote missing values
gene$fileSkipRows = 1          # one row of column labels
gene$fileSkipColumns = 1       # one column of row labels
gene$fileSliceSize = 2000      # read file in pieces of 2,000 rows
gene$LoadFile("../RNA_counts/n579_TPM_combat_adjusted_counts.txt")

cvrt = SlicedData$new()
cvrt$fileDelimiter = "\t"      # the TAB character
cvrt$fileOmitCharacters = "NA" # denote missing values
cvrt$fileSkipRows = 1          # one row of column labels
cvrt$fileSkipColumns = 1       # one column of row labels
cvrt$fileSliceSize = 2000      # read file in pieces of 2,000 rows
cvrt$LoadFile("../snp_array/covariates_and_phenotype_files/covariates_with_disease_status.txt") # covariates.txt
```

```{r}
#library(dplyr)
top.gene <- "ENSG00000155761.14"
top.snp <- "rs4880848" # rs2298746

genotypes <- sliced.data.row(snps, top.snp)
genotypes <- factor(genotypes, levels=0:2, labels=c("AA","AB","BB"))
expression.levels <- sliced.data.row(gene, top.gene)
gene_name <- subset(genes, gene_id == top.gene)
#pdf(file.path(out.dir, "top-snp-gene.pdf"))
boxplot(expression.levels ~ genotypes,
        main=paste(gene_name$gene_name, ": ", top.snp),
        ylim=range(expression.levels),
        outline=F)
stripchart(expression.levels ~ genotypes,
           method="jitter", add=TRUE, vertical=TRUE,
           col=c("blue","orange","black"), pch=19)
#dev.off()
```
```{r}
genotypes <- sliced.data.row(snps, top.snp)
genotypes <- factor(genotypes, levels=0:2, labels=c("CC","CT","TT"))
expression.levels <- sliced.data.row(gene, top.gene)
gene_expr <- as.data.frame(expression.levels)
snp <- as.data.frame(genotypes)
plot_data <- cbind(snp, gene_expr)
colnames(plot_data) = c("snp", "gene_expr")
plot_data = as.data.frame(plot_data)
plot_data$snp = as.factor(plot_data$snp)
head(plot_data)

FDR <- 4.720542e-148

lm_top = lm(plot_data[,"gene_expr"] ~ as.numeric(plot_data[,"snp"]))
summary(lm_top)

plot_data$IID <- rownames(plot_data)
df <- merge(plot_data, meta, by = "IID")
ggplot(df, aes(x = snp, y = log2(gene_expr), fill = df$TYPE)) +geom_violin() + geom_boxplot() +theme_bw()

```

```{r}
#look_up <- subset(eqtls_gene, gene_name == "RBP1")
top.snp <- "rs16943227" 
top.gene <- "ENSG00000204577.11"
# Assuming sliced.data.row function and data are available
genotypes <- sliced.data.row(snps, top.snp)
factor(genotypes)
genotypes <- factor(genotypes, levels=0:2, labels=c("0","1","2"))
genotypes <- factor(genotypes, levels=0:1, labels=c("0","1"))

expression.levels <- sliced.data.row(gene, top.gene)
gene_expr <- as.data.frame(expression.levels)
snp <- as.data.frame(genotypes)
plot_data <- cbind(snp, gene_expr)
colnames(plot_data) = c("snp", "gene_expr")
plot_data = as.data.frame(plot_data)
plot_data$snp = as.factor(plot_data$snp)

# Extracting p-value
goi <- subset(eqtls_gene, SNP == top.snp & gene == top.gene)

# Merge with meta data
plot_data$IID <- rownames(plot_data)
df <- merge(plot_data, meta, by = "IID")

# Count the number of samples for each genotype
sample_counts <- df %>%
  group_by(snp) %>%
  summarize(count = n())

sample_counts <- na.omit(sample_counts)
factor(df$snp)
# Add sample count labels to the x-axis
df$snp <- factor(df$snp, levels = c("0", "1", "2"),
                 labels = c(paste0("0 (", sample_counts$count[sample_counts$snp == "0"], ")"),
                            paste0("1 (", sample_counts$count[sample_counts$snp == "1"], ")"),
                            paste0("2 (", sample_counts$count[sample_counts$snp == "2"], ")")))

df$snp <- factor(df$snp, levels = c("0", "1"),
                 labels = c(paste0("0 (", sample_counts$count[sample_counts$snp == "0"], ")"),
                            paste0("1 (", sample_counts$count[sample_counts$snp == "1"], ")")))


df$TYPE <- factor(df$TYP, levels = c("CONTROL", "PA", "AD", "LBD"))
df$sex_inferred <- factor(df$sex_inferred)
# Define custom colors for each disease type
TypeColors <- c("#4682B4", "#B4AF46", "#B4464B", "gray35")


df <- df[!is.na(df$snp), ]
# Plot
ggplot(df, aes(x = snp, y = log2(gene_expr), fill = TYPE)) +
  geom_violin(position = position_dodge(width = 0.5), alpha = 0.6, draw_quantiles = NULL) +  # Adjust positioning and transparency
  geom_boxplot(width = 0.2, position = position_dodge(width = 0.5), outlier.shape = NA, alpha = 0.3) +  # Adjust width and position
  scale_fill_manual(values = TypeColors) +  # Apply custom colors
  theme_classic() +
  labs(x = paste0("FDR: ", format(goi$FDR, digits = 3, scientific = TRUE))) +
  theme(axis.title.x = element_text(color = "red")) +
  ggtitle(paste0(goi$gene_name, ": ", top.snp))
```




