---
title: "Plot eQTLs"
author: "Kimberly Olney, Ph.D."
date: "07/22/2024"
output:
  html_document:
    df_print: paged
  pdf_document: default
params:
  args: myarg
---
# Setup
```{r setup}
knitr::opts_knit$set(root.dir = ".")
library(dplyr)
library(data.table)
library(MatrixEQTL)
library(ggplot2)
```

# Read in MatrixEQTL data
```{r}
me <- fread("test")
genotype_data <- fread("../snp_array/Filtered_n598_CWOW_genotype_data.txt")
expression_data <- fread("../RNA_counts/n598_counts_filtered.txt")
meta <- fread("../metadata/filtered_metadata.txt")
gene_info <- fread("../RNA_counts/genes_filtered.txt")
gene_info$gene <- gene_info$gene_name
```

```{r}
snp_gene_info <- merge(me, gene_info, by = "gene")
genotype_data$SNP <- genotype_data$V1 
genotype_data$V1 <- NULL

geno <- melt(genotype_data, id.vars = "SNP")
```


```{r}
top.snp <- "rs17884850_A"
SERPINA3 <- subset(me, gene == "APOE")
#GPR3_geno <- subset(genotype_data, snp == "rs2230149_G")
#ahello <- melt(GPR3_geno)


#rownames(genotype_data) -> genotype_data$snp
#df <- melt(genotype_data)
top.gene <- "APOE"
rownames(genotype_data) <- genotype_data$V1
genotype_data$V1 <- NULL
genotype_data$snp <- rownames(genotype_data)
#genotypes <- sliced.data.row(genotype_data, top.snp)
snpoi <- subset(genotype_data, snp == top.snp) 
genotypes <- factor(snpoi, levels=0:2, labels=c("0","1","2"))
#expression.levels <- sliced.data.row(expression_data, top.gene)
expoi <- subset(expression_data, V1 == top.gene)
expoi$V1 <- NULL
df <- as.data.frame(expoi)
mdf <- melt(df)
#genotypes
geno <- as.data.frame(genotypes) 
geno_df <- geno[rownames(geno) != "snp", ]

#pdf(file.path(out.dir, "top-snp-gene.pdf"))
boxplot(mdf$value ~ geno_df,
        main=paste(top.snp,top.gene),
        ylim=range(expoi),
        outline=F)
stripchart(mdf$value ~ geno_df,
           method="jitter", add=TRUE, vertical=TRUE,
           col=c("blue","orange","black"), pch=19) 




```

```{r}
sig_eqtl <- subset(eqlts, eqlts$FDR < 0.00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000001)

sig_eqtl$pvalue
# Create the plot
library(ggplot2)
ggplot2eqtl_plot <- ggplot(sig_eqtl, aes(x = snps, y = gene, color = -log10(pvalue))) +
  geom_point() +
  scale_color_gradient(low = "blue", high = "red") +
  labs(title = "eQTL Plot",
       x = "SNP",
       y = "Gene",
       color = "-log10(p-value)") +
  theme_minimal()

# Display the plot
print(ggplot2eqtl_plot)

# Customize the plot
ggplot2eqtl_plot <- ggplot2eqtl_plot +
  theme(axis.text.x = element_text(angle = 90, hjust = 1),
        axis.text.y = element_text(angle = 0, hjust = 1),
        plot.title = element_text(hjust = 0.5)) +
  guides(color = guide_colorbar(barwidth = 10, barheight = 1))

# Display the customized plot
print(ggplot2eqtl_plot)


genotype_data <- fread("../snp_array/Filtered_n598_CWOW_genotype_data.txt")
expression_data <- fread("../RNA_counts/n598_counts_filtered.txt")
```

```{r}
# Load necessary libraries
library(dplyr)
library(ggplot2)

# Sample genotype data (replace with your actual data)
genotype_data <- data.frame(
  SampleID = c("MC00092", "MC00094", "MC00118", "MC00151", "MC00190", "MC01163"),
  Genotype = c(1, 0, 2, 0, 2, 0) # Add your actual genotype data here
)

# Sample expression data (replace with your actual data)
expression_data <- data.frame(
  SampleID = c("MC00092", "MC00094", "MC00118", "MC00151", "MC00190", "MC01163"),
  Expression = c(0.182, 0.207, 0.041, 0.300, 0.054, 0.057) # Add your actual expression data here
)

# Merge data by SampleID
merged_data <- inner_join(genotype_data, expression_data, by = "SampleID")

# Create the plot
ggplot(merged_data, aes(x = Genotype, y = Expression)) +
  geom_point() +
  labs(title = "eQTL Plot",
       x = "Genotype (0: Homozygous Reference, 1: Heterozygous, 2: Homozygous Alternate)",
       y = "Expression Level") +
  theme_minimal()

https://github.com/perishky/matrixeqtl-tutorial
```
