---
title: "plot GWAS"
author: "Kimberly Olney, Ph.D."
date: "March 2025"
output:
  html_document:
    df_print: paged
  pdf_document: default
params:
  args: myarg
---
# Setup
```{r setup}
knitr::opts_knit$set(root.dir = ".")
setwd("/tgen_labs/jfryer/kolney/LBD_CWOW/QTL/LBD_CWOW_QTL/scripts/")
source(here::here("/tgen_labs/jfryer/kolney/LBD_CWOW/QTL/LBD_CWOW_QTL/scripts/", "file_paths_and_colours.R"))
library(ggmanh)
```

# Read SNP annotation
```{r omni}
snp_anno <- fread("../snp_array/reference/NCBI_SNP_positions_with_gene_hg19ToHg38_lift_over.tsv", header = TRUE,  fill=TRUE)
snp_anno$gene <- snp_anno$Gene # create a gene column 
snp_anno$SNP <- snp_anno$names # create a SNP column for downstream matching 
```

# Cing LB
### GWAS 
```{r read_gwas_results}
# Read in GWAS file 
gwas <- fread("../snp_array/associations_with_imputed_snps/cingLBD_association_imputed.assoc.linear")

# Combine with snp annotation file and reformat 
gwas_snp <- merge(gwas, snp_anno, by = "SNP")
chr_labels <- c(1:22, "X")
```

### ggmanh
```{r}
gwas_snp$CHR[gwas_snp$CHR == 23] <- "X"
gwas_snp$CHR <- factor(gwas_snp$CHR, levels = chr_labels)
chr_colors <- c("navyblue", "skyblue2")

gwas_snp$label <-  ifelse(gwas_snp$P < 0.00001, gwas_snp$gene, NA)

g <- manhattan_plot(x = gwas_snp, pval.colname = "P", chr.colname = "CHR", pos.colname = "BP", plot.title = "GWAS Cingulate Lewy body count", y.label = "-log10(p-value)", chr.col = chr_colors, label.colname = "label", label.font.size = 2.75, highlight.colname = "label", highlight.col = c("black")) 

g
saveToPDF("../results/gwas/manhattan_plot_Cingulate_LB_count_imputed.pdf", width = 8, height = 4.5)
```

# Thal
### GWAS 
```{r read_gwas_results}
# Read in GWAS file 
gwas <- fread("../snp_array/associations_with_imputed_snps/Thal_association_imputed.assoc.linear")

# Combine with snp annotation file and reformat 
gwas_snp <- merge(gwas, snp_anno, by = "SNP")
levels(factor(gwas_snp$CHR))
chr_labels <- c(1:22, "X")
```

### ggmanh
```{r}
gwas_snp$CHR[gwas_snp$CHR == 23] <- "X"
gwas_snp$CHR <- factor(gwas_snp$CHR, levels = chr_labels)
chr_colors <- c("navyblue", "skyblue2")

gwas_snp$label <-  ifelse(gwas_snp$P < 0.00001, gwas_snp$gene, NA)

g <- manhattan_plot(x = gwas_snp, pval.colname = "P", chr.colname = "CHR", pos.colname = "BP", plot.title = "GWAS Thal amyloid phase", y.label = "-log10(p-value)", chr.col = chr_colors, label.colname = "label", label.font.size = 2.75, highlight.colname = "label", highlight.col = c("black")) 
g
saveToPDF("../results/gwas/manhattan_plot_Thal_amyloid_phase_imputed.pdf", width = 8, height = 4.5)
```

# Braak
### GWAS 
```{r read_gwas_results}
# Read in GWAS file 
gwas <- fread("../snp_array/associations_with_imputed_snps/Braak_association_imputed.assoc.linear")

# Combine with snp annotation file and reformat 
gwas_snp <- merge(gwas, snp_anno, by = "SNP")
levels(factor(gwas_snp$CHR))
chr_labels <- c(1:22, "X")
```

### ggmanh
```{r}
gwas_snp$CHR[gwas_snp$CHR == 23] <- "X"
gwas_snp$CHR <- factor(gwas_snp$CHR, levels = chr_labels)
chr_colors <- c("navyblue", "skyblue2")

gwas_snp$label <-  ifelse(gwas_snp$P < 0.00001, gwas_snp$gene, NA)

g <- manhattan_plot(x = gwas_snp, pval.colname = "P", chr.colname = "CHR", pos.colname = "BP", plot.title = "GWAS Braak NFT stage", y.label = "-log10(p-value)", chr.col = chr_colors, label.colname = "label", label.font.size = 2.75, highlight.colname = "label", highlight.col = c("black")) 
g
saveToPDF("../results/gwas/manhattan_plot_Braak_NFT_stage_imputed.pdf", width = 8, height = 4.5)
```


# Manhattan
```{r manhattan_default}
# Manhattan plot and show significant SNPs
sig <- subset(gwas_snp, P < 0.00001)
snpsOfInterest <- sig$SNP
manhattan(gwas_snp, highlight = snpsOfInterest)


# Show only chromosome 19 
manhattan(subset(gwas_snp, CHR == 19), main = "Chr 19")
# Show only X chromosome
manhattan(subset(gwas_snp, CHR == 23), main = "Chr X")
```

# QQplot
```{r qqplot_pvalue}
qq(gwas_snp$P, main = "Q-Q plot of GWAS p-values", xlim = c(0, 7), ylim = c(0,
    12), pch = 18, col = "blue4", cex = 1.5, las = 1)
```
